src/truss.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+
+  Copyright 2017 The Johns Hopkins University Applied Physics Laboratory LLC. All Rights Reserved.
+
+  Truss algorithm for cohesive subgroups.
+
+  Author: Alex Perrone
+  Date: 2017-08-03
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
+  02110-1301 USA
+
+*/
+
+#include "truss.h"
+#include <iostream>
+#include <vector>
+#include <unordered_set>
+
+using namespace std;
+
+/**
+ * \function trussness
+ * \brief Find the "trussness" for every edge in the graph. This
+ * indicates the highest k-truss that the edge occurs in.
+ *
+ * \param graph The input graph.
+ * \param truss Pointer to initialized vector of truss values that will
+ * indicate the highest k-truss each edge occurs in. It will be resized as
+ * needed.
+ *
+ * \example examples/simple/igraph_truss.cpp
+ */
+int igraph_truss(const igraph_t* graph, igraph_vector_int_t* truss){
+  igraph_vector_int_t triangles, support;
+  igraph_vector_t eid, unpacked_triangles;
+
+  // List the triangles as vertex triplets.
+  igraph_vector_int_init(&triangles, 0);  // will be resized
+  igraph_list_triangles(graph, &triangles);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &triangles);
+
+  // Unpack the triangles from vertex list to edge list.
+  long int num_triangles = igraph_vector_int_size(&triangles);
+  igraph_vector_init(&unpacked_triangles, 2 * num_triangles);
+  unpack(&triangles, &unpacked_triangles);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &unpacked_triangles);
+  igraph_vector_int_destroy(&triangles);
+  IGRAPH_FINALLY_CLEAN(1);
```

Comment:
Here we have a problem. This line discards the last entry from igraph's "finally stack" which keeps track of the memory chunks that were allocated during the execution of a function. However, the last entry in the finally stack at this point is not `triangles` (which we have just deallocated) but `unpacked_triangles`. If any error would happen after this point and we added `IGRAPH_CHECK()` calls everywhere, igraph would attempt to free `triangles` again (because it still stayed in the finally stack) and leak `unpacked_triangles` (which is not in the finally stack any more).

The solution is typically to allocate all the vectors that you would need right at the beginning of the function, with zero length, then put them in the finally stack in the order they will be deallocated during the execution. Then, whenever you figure out what the real size of a vector should be, you can call `IGRAPH_CHECK(igraph_vector_resize(...))` or `IGRAPH_CHECK(igraph_vector_int_resize(...))` to grow the vector to the right size.